### PR TITLE
Fix undefined steps for partial workflow

### PR DESCRIPTION
**Pull request by emcfarlane**
Have an error report of: `Error: Cannot read properties of undefined (reading 'status')`. I believe it's from the `steps[key].status == Status.Failed` if the value is in the object but undefined. Removed the class for an object that is allowed to be partial and added more `?.` guards for field access.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -55,8 +54,8 @@ async function main() {
   // NB: Write empties the buffer must be after the comment.
   await summary.write();
   // Finally, set the status of the action.
-  for (const key of Object.keys(steps)) {
-    if (steps[key].status == Status.Failed) {
+  for (const [key, value] of Object.entries(steps)) {
```
Comment (emcfarlane) [inline src/main.ts:57]: The `Object.entries` looses the type and goes to `string, any`.

Comment (paul-sachs) [reply src/main.ts:57]: Yes, unfortunately typescript is representing javascript correctly here. Steps could contain other things. You could help it be explicit with something like:

```ts
 for (const [key, value] of Object.entries(steps) as [
    keyof Steps,
    Result | undefined
  ][]) {
```

Will suffice to make TS happy.

Comment (emcfarlane) [reply src/main.ts]: Thanks! Added the type sig, but still don't get an error on `value.status` vs `value?.status`. Seems like the value doesn't respect the undefined or it makes no diff to the type check here?

Comment (paul-sachs) [reply src/main.ts]: Hmm, that is weird, haven't seen it such that it coerced. Might be a TS option that I'm missing somewhere.

**Review by nicksnyder — APPROVED** [review]

**Pull request merged**